### PR TITLE
feat: added GetAllFeatureVariables API.

### PR DESCRIPTION
**Pull request by yasirfolio3**
## Summary
- added GetAllFeatureVariables public API.

## Test Plan
- FSC should pass.
- Added Unit tests.

### CODE REVIEW

File: Sources/Optimizely/OptimizelyError.swift [hunk 1]
```diff
@@ -33,6 +33,7 @@ public enum OptimizelyError: Error {
     case variationUnknown(_ userId: String, _ key: String)
     case variableKeyInvalid(_ varKey: String, _ feature: String)
     case variableValueInvalid(_ key: String)
+    case variableTypeInvalid(_ type: String)
```
Comment (jaeopt) [inline Sources/Optimizely/OptimizelyError.swift:36]: Let's remove this error for forward compatibility

File: Sources/Optimizely/OptimizelyClient.swift [hunk 2]
```diff
@@ -591,12 +589,94 @@ open class OptimizelyClient: NSObject {
...
+                    break
+                }
+            } else {
+                logger.i(.variableTypeInvalid(v.type))
+                isValidType = false
+            }
+
+            if let value = valueParsed, isValidType {
+                variableMap[v.key] = value
+                sendDecisionNotification(decisionType: .featureVariable,
```
Comment (jaeopt) [inline Sources/Optimizely/OptimizelyClient.swift:658]: We're supposed to send only one notification for getAllFeatureVaraibles. See FJSON TDD spec for the new notification type.

File: Sources/Optimizely/OptimizelyClient.swift [hunk 3]
```diff
@@ -591,12 +589,94 @@ open class OptimizelyClient: NSObject {
...
+        for (_, v) in featureFlag.variablesMap {
+            var featureValue = v.value
+            if enabled, let variable = decision?.variation?.getVariable(id: v.id) {
+                featureValue = variable.value
+            }
+            
+            var valueParsed: Any? = featureValue
+            var isValidType = true
+            
+            if let valueType = Constants.VariableValueType(rawValue: v.type) {
```
Comment (jaeopt) [inline Sources/Optimizely/OptimizelyClient.swift:634]: I like this enum for types. This enum may cause an issue when we see unknown types (forward compatibility). Can we have some tests with "future" variable types?

Comment (yasirfolio3) [reply Sources/Optimizely/OptimizelyClient.swift:634]: for unknown/future types, the value is still be mapped in the variableMap as string, i have now added an additional test for this.

File: Tests/OptimizelyTests-APIs/OptimizelyClientTests_Valid.swift [hunk 4]
```diff
@@ -158,6 +158,39 @@ class OptimizelyClientTests_Valid: XCTestCase {
...
+        XCTAssert(variablesMap[kVariableKeyInt] as! Int == 43)
+        XCTAssert(variablesMap[kVariableKeyDouble] as! Double == 4.3)
+        XCTAssert((variablesMap[kVariableKeyJSON] as! [String: Any])["value"] as! Int == 2)
+    }
+    
+    func testGetAllFeatureVariablesFeatureDisabled() {
+        self.optimizely.config?.allExperiments[0].variations[0].featureEnabled = false
+        let optimizelyJSON = try? self.optimizely.getAllFeatureVariables(featureKey: kFeatureKey,
+                                                                         userId: kUserId)
+        let variablesMap = optimizelyJSON!.toMap()
```
Comment (jaeopt) [inline Tests/OptimizelyTests-APIs/OptimizelyClientTests_Valid.swift:186]: Let's add the variable settings here too to make sure disbled feature variable ignore them:

        config?.allExperiments[0].variations[0].variables?.append(contentsOf:[
            Variable(id: "2687470095", value: "43"),
            Variable(id: "2689280165", value: "4.3"),
            Variable(id: "2689660112", value: "false"),
            Variable(id: "2696150066", value: "f_foo"),
            Variable(id: "2696150067", value: "{\"value\":2}")
            ]
        )

**Comment by jaeopt** [comment]
Why is this still in Draft? It looks like it's ready for merge.

**Comment by yasirfolio3** [comment]
> Why is this still in Draft? It looks like it's ready for merge.

We were waiting for featureVariableJSON PR to be merged first, marking it ready for review now after rebasing it with master.

**Comment by yasirfolio3** [comment]
> LGTM

Can you please merge this aswell?

**Comment by jaeopt** [comment]
Need approval from Pawel. I'll merge them after that.

**Pull request merged**